### DISCUSSION TITLE
Unit Testing

### DISCUSSION THREAD

**Discussion by prithvisinha** [MarimerLLC/csla#1656] [Questions]
Hi,
I have been a CSLA user since v3.0 and have used it upto v4.0 but have been not in touch for the past 1-2 years. 
Coming back to my query, I have download the v5.2 and am able to run the projecttracker sample.
But I am unable to unit test the business objects (projectedit, projectlist). 

What is the right approach to do so and what methods is everyone else using.
Any sample code would be a great help.

Thanks
Prithvi

**Comment by rockfordlhotka** [comment]
The answer depends (can you tell I work in consulting? 😉) on exactly what you mean by "unit testing".

"Real" unit testing generally won't test domain types, because you should be testing your business rule types, and _maybe_ testing to ensure that the right rules are being associated with various properties of your domain class.

Determining whether a rule is associated with a property can be done without invoking the properties, because you can interrogate the rules engine to see which rules are associated with each property.

What most people _actually_ do is some level of "integration testing", where they create an instance of a domain type, reading/writing properties, and checking the result. This actually is a _long way_ from unit testing, because this sort of testing invokes the data portal, probably invokes persistence, and also invokes business rules - but within the context of simulated UI binding.

The way you implement _unit testing_ is different from _integration testing_. And even within integration testing, there are various levels of abstraction/complexity you can choose - ranging from high level testing, down to very sophisticated mocking of things like the data portal.

**Reply by prithvisinha** [reply]
Well, actually even I come in "most people". 

What I really meant was that, 
_How can I create domain types, read / write properties and check the results._

**Reply by prithvisinha** [reply]
A few years back, in CSLA version 4.5 there was a test library in projecttracker which could test the domain types. Mere mortals like us are still trying to wrap my head around CSLA and depend upon this mixed-up Unit-Integration testing. 

I am trying to learn the new dal  [inject] syntax. 
Kindly share any sample code.

**Reply by rockfordlhotka** [reply]
I meant no disrespect @prithvisinha - it is just that we put a lot of work into supporting unit testing, in addition to the original support for integration testing, and I wanted to make sure I understood what you were after.

I don't recall there being a unit test project in ProjectTracker, though there might have been at one point.

Generally, the process looks like this:

1. Have a mock DAL you can switch in for your real DAL (or a way to use your real DAL against a known set of data that's reset before each test run)
1. Configure the use of your mock DAL in the test init method(s)
1. Write tests that create/fetch an object graph in a known state
   1. Manipulate the known state to trigger selected rules
   1. Assert that the rules are broken or alter state as expected

That's pretty much it. Of course this can get very complex if your rules are complex, or if your object graph is complex and needs a lot of code to get it into a specific state for testing.

**Reply by rockfordlhotka** [reply]
Here's a quirky example from an app I'm building to manage a table-top role playing game. This is a test that gets a `Character` domain object, manipulates some of its state (by using other domain objects), and then checks to see if the rules ran to create the expected result values.

```c#
    [TestMethod]
    public void SingleAttributeBonus()
    {
      var c = DataPortal.Create<Character>("");
      c.AttributeList.Where(r => r.Name == "STR").First().Value = 7;
      var cs = new CharacterSkill
      {
        PrimarySkill = "STR",
        Level = 3
      };
      var s = DataPortal.FetchChild<Skill>(cs);
      c.Skills.Add(s);
      Assert.AreEqual("STR", s.PrimaryAttribute);
      Assert.AreEqual(3, s.Level);
      Assert.AreEqual(-2, s.Bonus);
      Assert.AreEqual(5, s.AbilityScore);
    }
```

**Reply by prithvisinha** [reply]
Thanks. I know you didnt mean any disrespect and I also intended my statement to be funny. 

By the way, while waiting for any solution, I went through the blazor book.

But, I still dont know whether it is the right approach

**Reply by prithvisinha** [reply]
> I don't recall there being a unit test project in ProjectTracker, though there might have been at one point.

Actually, I have the source code for v 4.5.701 which has the unit tests.

[ProjectTracker.Test.Library.zip](https://github.com/MarimerLLC/csla/files/4732107/ProjectTracker.Test.Library.zip)

**Reply by prithvisinha** [reply]
> 2. Configure the use of your mock DAL in the test init method(s)

This part is proving to be tough.
The injected dal in my fetch methods show up as null

**Reply by rockfordlhotka** [reply]
The standard mstest project template doesn't set up dependency injection during startup, so you need to do that. At the _very least_ you need to create and build a `ServicesCollection` and provider.

iirc, this is done with a method like this in _exactly one_ of your `TestClass` classes.

```c#
    [AssemblyInitialize]
    public void GlobalSetup(TestContext context)
    {
      var services = new ServiceCollection();
      // services.AddTransient or AddScope for your DAL
      services.AddCsla();
    }
```

**Reply by prithvisinha** [reply]
Even I am using the same approach but still no joy. 

```
 [TestFixture]
    public class BaseClass
    {
        public static IServiceProvider ServiceProvider { get; set; }

        [OneTimeSetUp]
        public static void AssemblyInit()
        {
            // basically a "app builder" implementation
            var serviceCollection = new ServiceCollection();
            serviceCollection.AddCsla();
            serviceCollection.AddDalMock();

            serviceCollection.AddScoped((c) =>
                Startup.LoadAppConfiguration(Array.Empty<string>()));
            var startup = ActivatorUtilities.CreateInstance<Startup>(
                serviceCollection.BuildServiceProvider(), Array.Empty<object>());
            startup.ConfigureServices(serviceCollection);
            ServiceProvider = serviceCollection.BuildServiceProvider();
            startup.Configure();
        }
    }

public class Startup
    {
        /// <summary>
        /// Load app configuration via ConfigurationBuilder
        /// </summary>
        /// <returns>App configuration</returns>
        public static IConfiguration LoadAppConfiguration(string[] args)   {
            return new ConfigurationBuilder().AddInMemoryCollection().Build();
        }

        /// <summary>
        /// Creates instance of type, getting services via DI
        /// </summary>
        /// <param name="config">App configuration</param>
        public Startup(IConfiguration config)    {
            Configuration = config;
        }

        public IConfiguration Configuration { get; }

        /// <summary>
        /// This method gets called by the runtime. 
        /// Use this method to add services to the container.
        /// </summary>
        /// <param name="services">Service collection</param>
        public void ConfigureServices(IServiceCollection services)
        {   }

        /// <summary>
        /// This method gets called by the runtime. 
        /// Use this method to configure the app.
        /// </summary>
        public void Configure()      {
            CslaConfiguration.Configure().PropertyChangedMode(Csla.ApplicationContext.PropertyChangedModes.Windows)
                .DataPortal().DefaultProxy(typeof(Csla.DataPortalClient.HttpProxy),
                    "http://localhost:8040/api/dataportal/");
        }
    }
```

**Reply by rockfordlhotka** [reply]
Have you set a breakpoint in your `AssemblyInit` method to make sure it is being called? I think that method is supposed to accept a `TestContext` parameter?

**Reply by prithvisinha** [reply]
> Have you set a breakpoint in your `AssemblyInit` method to make sure it is being called? I think that method is supposed to accept a `TestContext` parameter?

It is being called.
I am using nunit and a method with [OneTimeSetUp] attribute cant have any parameters.

Fetch method is not getting called but strangely if it run the whole project everything works smoothly

![image](https://user-images.githubusercontent.com/66350767/83802961-5309a600-a6c9-11ea-802b-c0bab7723357.png)

**Reply by rockfordlhotka** [reply]
You might try explicitly setting the service provider in CSLA so it can use the provider.

```c#
      Csla.Configuration.CslaConfiguration.Configure()
        .ServiceProviderScope(ServiceProvider);
```

**Reply by prithvisinha** [reply]
Well, I finally got it working.
Removed all csla.configuration parts from code. I am still trying to understand as to how it works
@rockfordlhotka your views on this are needed

I am describing the steps.
1) Add a new **Nunit .net core project** to the projecttracker solution
2) Add **Microsoft.AspNetCore.TestHost** to the UnitTest Project
3) Create a new class and add this code.

`using System.Net.Http;`
`using ProjectTracker.AppServerCore;`
`using Microsoft.AspNetCore.Hosting;`
`using Microsoft.AspNetCore.TestHost;`
`using NUnit.Framework;`

`[SetUpFixture]`
`public class RunOnce`
`{`
    `[OneTimeSetUp]`
    `public static void AssemblyInit()`
    `{`
        `var server = new TestAppServer();`
    `}`
`}`

`public class TestAppServer`
`{`
    `private readonly HttpClient _client;`

    public TestAppServer()
    {
        var server = new TestServer(new WebHostBuilder()
            .UseEnvironment("Development")
            .UseStartup<Startup>());
        _client = server.CreateClient();
    }
}`

_The above code shouldn't be in any namespace._

**Comment by Chicagoan2016** [comment]
@rockfordlhotka , agree about the integration testing part, my unit tests are technically integration tests.
